Refactor conversions from LazyMetadataValue into abstract base class (BUKKIT-3624)

crast opened the pull request:
Implementing MetadataValue interface is significant work due to having to provide a large amount of conversion stub methods. This adds a new optional abstract base class to aid in implementation, and has LazyMetadataValue extend this.

Includes comprehensive unit tests including a sample adapted class, and all existing metadata tests pass as well.

This supercedes an older pull request https://github.com/Bukkit/Bukkit/pull/580

Also see associated Atlassian ticket: https://bukkit.atlassian.net/browse/BUKKIT-3624

crast committed:
Refactor conversions from LazyMetadataValue into abstract base class

Implementing MetadataValue interface is significant work due to having
to provide a large amount of conversion stub method. This adds a new
optional abstract base class to aid in implementation.

Includes comprehensive unit tests including a sample adapter class, and
all existing metadata tests pass.

Wolvereness commented:
The unit test relies on specific implementation calls (the value++ thing). I recommend a set of tests and expected values clearly listed. Some of the testing standards I've been using can be seen in https://github.com/Bukkit/Bukkit/blob/master/src/test/java/org/bukkit/ColorTest.java. One of those is assertThat, and the use of matcher wrapping.

crast committed:
Clean up whitespace.

crast committed:
Substantially more comprehensive unit tests.

Check all the interesting implementation details in metadatavalues
which were never tested before, as well as making sure we document
things thoroughly.

crast commented:
@Wolvereness 
using `assertThat` and matchers has one flaw: they autobox primitive types. While this on the surface seems fine, the specific use case here is testing conversions involving all primitive types, which `assertEquals` can handle due to it being polymorphic on those types. Furthermore, the issue that assertThat is meant to address (more detail in assertion errors) doesn't apply for this use case.

amaranth commented:
Pulled, thanks.
